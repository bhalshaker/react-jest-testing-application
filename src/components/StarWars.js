import React, { useEffect, useState } from 'react';
import axios from 'axios';

const StarWars = () => {
  const [films, setFilms] = useState([]);
  const [error, setError] = useState(null);

  useEffect(() => {
    const source = axios.CancelToken.source();

    const fetchFilms = async () => {
      try {
        const { data } = await axios.get('https://swapi.dev/api/films/', {
          cancelToken: source.token,
        });
        setFilms(data.results);
      } catch (err) {
        if (axios.isCancel(err)) {
          console.log(`Request cancelled: ${err.message}`);
        } else {
          console.error(err);
          setError(err);
        }
      }
    };

    fetchFilms();

    return () => {
      source.cancel('Cancelling axios in cleanup');
    };
  }, []);

  if (error) {
    return (
      <section>
        <p class="error"> Error: {error.message} - Simulated API Response: </p>
        <h2>Star Wars: The Return of the Jedi</h2>
        <p>
          The Galactic Empire is on the brink of defeat. Rebel forces are
          gathering their strength for a final assault on the Empireâ€™s ultimate
          weapon, the second Death Star. Meanwhile, the fate of Han Solo rests
          in the hands of Princess Leia, Luke Skywalker, and the Ewoks on the
          forest moon of Endor.
        </p>
      </section>
    );
  }

  return (
    <section>
      <ul>
        {films.map(({ episode_id, title, opening_crawl }) => (
          <li key={episode_id}>
            <h2>{title}</h2>
            <p>{opening_crawl}</p>
          </li>
        ))}
      </ul>
    </section>
  );
};

export default StarWars;
