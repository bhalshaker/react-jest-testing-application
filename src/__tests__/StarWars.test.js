import React from 'react';
import { act, render, waitFor, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import StarWars from '../components/StarWars';

const server = setupServer(
  rest.get('https://swapi.dev/api/films/', (_req, res, ctx) => {
    return res(
      ctx.json({
        results: [
          {
            title: 'Star Wars: The Return of the Jedi',
            episode_id: 6,
            opening_crawl:
              'The Galactic Empire is on the brink of defeat. Rebel forces are gathering their strength for a final assault on the Empire’s ultimate weapon, the second Death Star. Meanwhile, the fate of Han Solo rests in the hands of Princess Leia, Luke Skywalker, and the Ewoks on the forest moon of Endor.',
          },
        ],
      })
    );
  })
);

beforeAll(() => server.listen());
afterAll(() => server.close());

describe('<StarWars/>', () => {
  it('displays film info returned from API call', async () => {
    await act(async () => {
      const component = render(<StarWars />);
      expect(component.baseElement).toMatchSnapshot('StarWars before data');

      // wait until what we expect to be on the page is actually there
      await waitFor(() =>
        component.getByText('Star Wars: The Return of the Jedi')
      );
      expect(component.baseElement).toMatchSnapshot('StarWars after data');

      // by now, all the text from the api should have been rendered
      expect(
        screen.getByText(
          'The Galactic Empire is on the brink of defeat. Rebel forces are gathering their strength for a final assault on the Empire’s ultimate weapon, the second Death Star. Meanwhile, the fate of Han Solo rests in the hands of Princess Leia, Luke Skywalker, and the Ewoks on the forest moon of Endor.'
        )
      ).toBeInTheDocument();
    });
  });
});
