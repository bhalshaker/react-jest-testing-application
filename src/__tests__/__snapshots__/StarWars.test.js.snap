// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarWars/> displays film info returned from API call: StarWars after data 1`] = `
<body>
  <div>
    <section>
      <ul>
        <li>
          <h2>
            Star Wars: The Return of the Jedi
          </h2>
          <p>
            The Galactic Empire is on the brink of defeat. Rebel forces are gathering their strength for a final assault on the Empireâ€™s ultimate weapon, the second Death Star. Meanwhile, the fate of Han Solo rests in the hands of Princess Leia, Luke Skywalker, and the Ewoks on the forest moon of Endor.
          </p>
        </li>
      </ul>
    </section>
  </div>
</body>
`;

exports[`<StarWars/> displays film info returned from API call: StarWars before data 1`] = `
<body>
  <div>
    <section>
      <ul />
    </section>
  </div>
</body>
`;
